sdk/eventgrid/eventgrid/src/consumer.ts

```diff
@@ -0,0 +1,170 @@
+   */
+  public async decodeEventGridEvents(encodedEvents: string): Promise<EventGridEvent<unknown>[]>;
+  /**
+   * Decodes events encoded in the Event Grid schema.
+   *
+   * @param encodedEvents an object representing a single event, encoded in the Event Grid schema.
+   */
+  public async decodeEventGridEvents(encodedEvents: object): Promise<EventGridEvent<unknown>[]>;
+  public async decodeEventGridEvents(
+    encodedEvents: string | object
+  ): Promise<EventGridEvent<unknown>[]> {
+    const decodedArray = parseAndWrap(encodedEvents);
+
+    const events: EventGridEvent<unknown>[] = [];
+
+    for (const o of decodedArray) {
+      if (typeof o !== "object") {
+        throw new TypeError("event is not an object");
+      }
+
+      validateRequiredStringProperty(o, "eventType");
+      validateRequiredStringProperty(o, "eventTime");
+      validateRequiredStringProperty(o, "id");
+      validateRequiredStringProperty(o, "subject");
+      validateRequiredStringProperty(o, "topic");
+      validateRequiredAnyProperty(o, "data");
+      validateRequiredStringProperty(o, "dataVersion");
+      validateRequiredStringProperty(o, "metadataVersion");
+
+      if (o.metadataVersion !== "1") {
+        throw new TypeError("event is not in the Event Grid schema");
+      }
+
+      const deserialized: EventGridEvent<any> = serializer.deserialize(EventGridEventMapper, o, "");
+
+      if (systemDecoders[deserialized.eventType]) {
+        deserialized.data = await systemDecoders[deserialized.eventType](deserialized.data);
+      } else if (this.customDecoders[deserialized.eventType]) {
+        deserialized.data = await this.customDecoders[deserialized.eventType](deserialized.data);
+      }
+
+      events.push(deserialized as EventGridEvent<unknown>);
+    }
+
+    return events;
+  }
+
+  /**
+   * Decodes events encoded in the Cloud Events 1.0 schema.
+   *
+   * @param encodedEvents the JSON encoded representation of either a single event or an array of
+   * events, encoded in the Cloud Events 1.0 Schema.
+   */
+  public async decodeCloudEvents(encodedEvents: string): Promise<CloudEvent<unknown>[]>;
+  /**
+   * Decodes events encoded in the Cloud Events 1.0 schema.
+   *
+   * @param encodedEvents an object representing a single event, encoded in the Cloud Events 1.0 schema.
+   */
+  public async decodeCloudEvents(encodedEvents: object): Promise<CloudEvent<unknown>[]>;
+  public async decodeCloudEvents(encodedEvents: string | object): Promise<CloudEvent<unknown>[]> {
+    const decodedArray = parseAndWrap(encodedEvents);
+
+    const events: CloudEvent<unknown>[] = [];
+
+    for (const o of decodedArray) {
+      if (typeof o !== "object") {
+        throw new TypeError("encoded event is not an object");
+      }
+
+      // Check that the required fields are present and of the correct type and the optional fields are missing
+      // or of the correct type.
+
+      validateRequiredStringProperty(o, "type");
```

Comment:
I'm also thinking you could make a meta-helper for each event type like `validateCloudEvent()` that could slim down the code in this method considerably